4124 workflow clean rollback

michbarsinai opened the pull request:
## Related Issues

- connects to #4124

## Pull Request Checklist

- [x] Unit [tests][] completed
- [ ] Integration [tests][]: None
- [ ] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [ ] [Documentation][docs] completed
- [x] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: https://github.com/IQSS/dataverse/tree/develop/scripts/database/upgrades
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/4.6.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

## General discussion
The  core issue here was the fact that when an `InReview` dataset is starting the pre-publication workflow, it should be locked twice: first for the in-review, then for the workflow. This way, if the workflow fails, the `InReview` lock stays.

To implement this, Datasets now have multiple locks. The API became simpler, though, because of the new method `dataset.isLockedFor( DatasetLock.Reason )`. Using this is an improvement over ` (dataset.getLock() != null && dataset.getLock().getReason == REASON`.

It's also a nice infrastructure for having multiple locks (e.g. a few PIs have to approve publication, multiple background processes...)

matthew-a-dunlap commented on src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java:
```diff
@@ -553,28 +547,18 @@ public DatasetLock addDatasetLock(Long datasetId, DatasetLock.Reason reason, Lon
 
     @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
-    public void removeDatasetLock(Long datasetId) {
+    public void removeDatasetLock(Long datasetId, DatasetLock.Reason aReason) {
```
It looks like with this implementation that multiple locks could be removed if they share the same reason. I am assuming that we are ensuring only one lock per reason is added, or that in all cases we want all locks of a reason to be removed. Even if one of these two are true, maybe this should be renamed to reflect what is happening with the filter.

matthew-a-dunlap commented:
Looks good to me aside from my one question!

michbarsinai commented:
Good catch, thanks @matthew-a-dunlap!
Method renamed to reflect that multiple locks might be removed. Also added some documentation and removed an unused and outdated method for returning lock info (`checkDatasetLockInfo`).

michbarsinai commented:
The build failure seems to arise from problems in the geotoolkit repo: https://travis-ci.org/IQSS/dataverse/builds/288195273#L1784

Need to try and rebuild after Monday noon.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/Dataset.java:
```diff
@@ -100,8 +100,8 @@
-    @OneToOne(mappedBy = "dataset", cascade = {CascadeType.REMOVE, CascadeType.MERGE, CascadeType.PERSIST}, orphanRemoval = true)
-    private DatasetLock datasetLock;
+    @OneToMany(mappedBy = "dataset", cascade = CascadeType.ALL, orphanRemoval = true)
+    private Set<DatasetLock> datasetLocks;
```
Ah ha! This change to support multiple locks is what @sekmiller and I believe we need for #4139 to fix a bug related to dataset leaving the in review state when the curator uploads a tabular file: https://github.com/IQSS/dataverse/issues/4139#issuecomment-337361146

pdurbin commented:
Closing in favor of pull request #4216.
